8366401: JCK test api/java_text/DecimalFormatSymbols/serial/InputTests.html fails after JDK-8363972

justin-curtis-lu opened the pull request:
This PR addresses a JCK test failure related to `DecimalFormatSymbols` de-serialization. While the current public API of DFS disallows a null locale, it was possible to set in the past. Thus, the `loadNumberData(locale)` call currently throws NPE when locale is null in the stream. The call should be guarded with a null check, such that if locale is null, then `lenientMinusSigns` defaults to `minusSignText`.

Defaulting the locale field when `null` to Locale.ROOT is also a reasonable solution, but I think that the current one is preferable as a user would not expect locale data related logic to occur if locale is `null`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8366401](https://bugs.openjdk.org/browse/JDK-8366401): JCK test api/java_text/DecimalFormatSymbols/serial/InputTests.html fails after JDK-8363972 (**Bug** - P2)


### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/27008/head:pull/27008` \
`$ git checkout pull/27008`

Update a local copy of the PR: \
`$ git checkout pull/27008` \
`$ git pull https://git.openjdk.org/jdk.git pull/27008/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 27008`

View PR using the GUI difftool: \
`$ git pr show -t 27008`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/27008.diff">https://git.openjdk.org/jdk/pull/27008.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/27008#issuecomment-3237816297)
</details>

naotoj reviewed (approved):
Since this is avoiding non-standard situations either solutions should be acceptable.

justin-curtis-lu commented:
@AlanBateman @naotoj  Right, we should not have to rely on the JCK tests to discover an issue, I've removed the `noreg-jck` label and added a serialization test. It checks all of the correct behavior based on the stream version, as well as other invariants. I also filed [JDK-8366733](https://bugs.openjdk.org/browse/JDK-8366733) becasue many of the existing serialization tests for DFS seem to be relics and can either be removed or are redundant with the newly added test.

justin-curtis-lu commented on test/jdk/java/text/Format/DecimalFormat/DFSSerializationTest.java:
```diff
@@ -0,0 +1,262 @@
+
+    // Utility to deserialize
+    private static DecimalFormatSymbols deSer(byte[] bytes) throws IOException, ClassNotFoundException {
+        try (ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
+             ObjectInputStream ois = new ObjectInputStream(byteArrayInputStream)) {
+            return (DecimalFormatSymbols) ois.readObject();
+        }
```
I did not swallow the error with `assertDoesNotThrow()` here becasue `disagreeingTextTest` expects to see an `InvalidObjectException.class`. Future tests/changes can easily check for expected de-serialization failures with this setup.

naotoj reviewed (approved):
LGTM

justin-curtis-lu commented:
/integrate
